### ISSUE TITLE
StaticReflectionParser is dumb, assumes Class::class references are class level

### ISSUE THREAD

**Issue by heddn** [doctrine/common#744]
If I use the below PHP class structure, then parsing the annotations will always return an empty class level annotation.  The current HEAD of https://www.drupal.org/project/migrate_source_csv is broken right now using this pattern. It is a sneaky breakage, because Unit tests pass. It is only when I run integration tests that is breaks.

See https://www.drupal.org/node/2818891

``` php
/**
 * Source for CSV.
 *
 * @MigrateSource(
 *   id = "csv"
 * )
 */
class CSV extends SourcePluginBase {

  protected $fileClass = '';

  /**
   * {@inheritdoc}
   */
  public function __construct(array $configuration, $plugin_id, $plugin_definition, MigrationInterface $migration) {
    $this->fileClass = empty($configuration['file_class']) ? CSVFileObject::class : $configuration['file_class'];
  }
```

**Comment by Ocramius** [comment]
I don't see an example in this issue: can you make one? Maybe a failing test?

**Comment by heddn** [comment]
I could get it to break by swapping

`$this->fileClass = empty($configuration['file_class']) ? 'Drupal\migrate_source_csv\CSVFileObject' : $configuration['file_class'];`

for 
`$this->fileClass = empty($configuration['file_class']) ? CSVFileObject::class : $configuration['file_class'];`

I'll add a patch to see if that reproduces it.

**Comment by heddn** [comment]
I've added https://www.drupal.org/node/2863902

**Comment by Ocramius** [comment]
Ah, so it's an actual parser failure. Is this using the latest version too? I have loads of `::class` usages, and never experienced this.

Also, please send tests and comments here - don't even know how the drupal issue system works :-\

**Comment by heddn** [comment]
I haven't had a chance to reproduce the issue directly in doctrine. But if I recall correctly, the key difference for me was that the  `::class` is in the `__construct()` call.

**Comment by ostrolucky** [comment]
I have successfully reproduced the issue (and fixed it). This happens every time optimization is off and there is some ::class used in a class

**Comment by Sam152** [comment]
I just had the same issue annotation parsing in Drupal. Tracked it down to the exact same place. I fixed it like this: #806. @ostrolucky fix did not seem to work if there was a doc comment on top of "::class".

**Comment by Ocramius** [comment]
Handled in #806
